### PR TITLE
Cmake cleanup

### PR DESCRIPTION
**Pull request by meshula**
This commit allows a full cmake build.

- consolidate options at the top level
- detect a "reasonable" place to install the python components
- restrict gcov coverage to platforms that support it (those built with gcc)
- use cmake build from setup,py on all platforms (cmake will call make appropriately, so the outcome is identical to how it was)
- update submodules (any/optional/etc) BEFORE installing them
- add ability to build OTIO libraries as static
- add optional debug postfix because windows release/debug builds are not compatible
- add platform appropriate rpaths (@loader_path on Mac, $ORIGIN on Linux)
- add ability for cmake to do an install into the python environment so that a cmake build automatically results in a functional python build
- more granular build control (turn python builds or c++ builds on or off independently)
- For IDEs, rename the .sln or .xcode file to OpenTimelineIO (was OPENTIMELINEIO_ROOT)

### CODE REVIEW

**Comment by meshula** [comment]
Related: https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/813

Before merging, I need to test pip's --target directive to make sure the target is properly transmitted to the cmake script.

**Comment by meshula** [comment]
Maybe related: https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/838 ~ If there is a more general PyInstaller issue, we could reopen 838 after this PR is merged.

**Comment by Tilix4** [comment]
Thanks! I confirm this will solve #813. I've tested the `--target` option and it works well on my windows.

Thanks, hope this will land soon :)

File: setup.py [hunk 1]
```diff
@@ -44,82 +44,98 @@ class _Ctx(object):
...
 
     if _ctx.cxx_coverage:
-        cmake_args += ['-DCXX_COVERAGE=1'] + cmake_args
+        cmake_args += ['-DOTIO_CXX_COVERAGE=1']
+
+    cmake_args = ['cmake', _ctx.source_dir] + cmake_args
+
+    print("------------------------------------")
+    print(cmake_args)
+    print("------------------------------------")
```
Comment (ssteinbach) [inline setup.py:107]: This looks like debug code

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,10 +1,201 @@
...
+    # to the Microsoft debug CRT.
+    set(OTIO_DEBUG_POSTFIX "d")
+endif()
+
+#------------------------------------------------------------------------------
+# Fetch or refresh submodules if requested
+#
+# fetching submodules does not work in Travis, so override the OTIO_AUTOMATIC_SUBMODULES option
+
+if (OTIO_AUTOMATIC_SUBMODULES AND NOT DEFINED ENV{TRAVIS})
```
Comment (ssteinbach) [inline CMakeLists.txt:155]: can you add a TODO note to rename the TRAVIS variable to something else?  We aren't using travis anymore for CI

File: setup.py [hunk 3]
```diff
@@ -166,61 +184,27 @@ def run(self):
...
-            )
-            if cmake_version < '3.1.0':
-                raise RuntimeError("CMake >= 3.1.0 is required on Windows")
+        ) and platform.system() == "Linux"
 
         self.build()
 
     def build(self):
-        _ctx.ext_dir = os.path.join(os.path.abspath(self.build_lib), "opentimelineio")
+        print("OTIO_build_ext::build is running cmake in the build location")
```
Comment (ssteinbach) [inline setup.py:192]: Is this debugging information?  Or do we want to keep this message

File: src/py-opentimelineio/opentimelineio-bindings/CMakeLists.txt [hunk 4]
```diff
@@ -11,12 +14,94 @@ pybind11_add_module(_otio
...
+
+install(TARGETS _otio 
+        ARCHIVE DESTINATION "${OTIO_RESOLVED_CXX_DYLIB_INSTALL_DIR}"
+        LIBRARY DESTINATION "${OTIO_RESOLVED_CXX_DYLIB_INSTALL_DIR}"
+        RUNTIME DESTINATION "${OTIO_RESOLVED_CXX_DYLIB_INSTALL_DIR}")
+
+#install(FILES 
+#        $<TARGET_FILE:opentime> 
+#        $<TARGET_FILE:opentimelineio> 
+#        DESTINATION "${OTIO_RESOLVED_PYTHON_INSTALL_DIR}/opentimelineio")
```
Comment (ssteinbach) [inline src/py-opentimelineio/opentimelineio-bindings/CMakeLists.txt:60]: dead code can be removed?

**Review by ssteinbach — CHANGES_REQUESTED** [review]
couple of really small questions around commenting and removing some debug statements

File: setup.py [hunk 5]
```diff
@@ -95,23 +89,29 @@ def compute_cmake_args():
...
-        cmake_args += ['-DCMAKE_BUILD_TYPE=' + cfg]
 
     if _ctx.cxx_coverage:
-        cmake_args += ['-DCXX_COVERAGE=1'] + cmake_args
+        cmake_args += ['-DOTIO_CXX_COVERAGE=1']
 
-    env = os.environ.copy()
+    print("------------------------------------")
+    print(cmake_args)
+    print("------------------------------------")
```
[the remark this answers is not in the document]
Comment (meshula) [reply setup.py]: removed

File: setup.py (hunk 3 above)
Comment (meshula) [reply setup.py:187]: That was leftover from learning how pip interacts with the various classes and functions in setup.py. Removing.

File: src/py-opentimelineio/opentimelineio-bindings/CMakeLists.txt (hunk 4 above)
Comment (meshula) [reply src/py-opentimelineio/opentimelineio-bindings/CMakeLists.txt:60]: left over from WIP in doing a complete python install from cmake. removed.

**Review by ssteinbach — APPROVED** [review]

**Pull request merged**